docs/ref-manual/cvl/expr.md

```diff
@@ -1,85 +1,289 @@
 Expressions
 ===========
 
-TODO: CVL expressions
+A CVL expression is anything that represents a value.  This page documents all
+possible expressions in CVL and explains how they are evaluated.
 
-TODO: organization
+```
+expr ::= literal
+       | unop expr
+       | expr binop expr
+       | "(" exprs ")"
+       | expr "?" expr ":" expr
+       | [ "forall" | "exists" ] type id "." expr
+
+       | expr "." id
+       | id [ "[" expr "]" { "[" expr "]" } ]
+       | id "(" types ")"
+
+       | [ "invoke" | "sinvoke" ]
+         [ id "." ] id
+         [ "@" ( "norevert" | "withrevert" | "dontsummarize" ]
+         "(" exprs ")"
+         [ "at" id ]
+
+       | expr "in" id
+
+
+literal ::= "true" | "false" | number | string
+
+unop  ::= "~" | "!" | "-"
+
+binop ::= "+" | "-" | "*" | "/" | "%" | "^"
+        | ">" | "<" | "==" | "!=" | ">=" | "<="
+        | "&" | "|" | "<<" | ">>" | "&&" | "||"
+        | "=>" | "<=>" | "xor" | ">>>"
+
+specials_fields ::=
+           | "block" "." [ "number" | "timestamp" ]
+           | "msg"   "." [ "address" | "sender" | "value" ]
+           | "tx"    "." [ "origin" ]
+           | "length"
+           | "selector" | "isPure" | "isView" | "numberOfArguments" | "isFallback"
+
+special_vars ::=
+           | "lastReverted" | "lastHasThrown"
+           | "lastStorage"
+           | "allContracts"
+           | "lastMsgSig"
```

Comment:
@shellygr allContracts and lastMsgSig, what are they and do we want to support them? maybe better remove from the docs

Reply:
We can try to remove them, see if all tests pass, and then remove from docs too

Reply:
They are currently marked as undocumented in the section of the document where they should be described; I think it makes sense to leave it that way until either someone figures out what they do or they are removed.